File: javascript/src/adapters/rasa.js
const fp = require('../fp');
const utils = require('../utils');
const Augment = require('../augment');

// As per https://rasa.com/docs/rasa/1.1.4/nlu/training-data-format/
module.exports = function generateTrainingDataset (chatl, options = {}) {
  const augment = new Augment(chatl, true);

  const getRealEntity = entityName => {
    const type = chatl.entities[entityName].props['type'];

    if (type && chatl.entities[type]) {
      return type;
    }

    return entityName;
  }

  const getRegexProp = entityName => chatl.entities[entityName].props['regex'];

  // For rasa, we need a map of synonyms -> value
  const synonymsLookup = fp.reduce((acc, synonyms, value) => 
    fp.append(...fp.map(s => ({ [s]: value }))(synonyms))(acc)
  )(augment.synonymsValues);

  const buildLookupTable = (acc, _, name) => {
    const entityName = getRealEntity(name);

    // Entity has regex feature, returns now
    if (getRegexProp(entityName)) {
      return acc;
    }

    return fp.append({
      name,
      elements: augment.getEntity(entityName).all(),
    })(acc);
  };

  const buildIntentExamples = (acc, intent, name) => {
    return fp.append(fp.map(sentence => {
      const entities = [];

      return {
        intent: name,
        text: fp.reduce((p, c) => {
          if (!utils.isEntity(c)) {
            return p + c.value;
          }

          const entityName = getRealEntity(c.value);
          const value = augment.getEntity(entityName).next(c.variant);

          entities.push({
            start: p.length,
            end: p.length + value.length,
            entity: c.value,
            value: synonymsLookup[value] || value, // Check if its a synonym here
          });

          return p + value;
        }, '')(sentence),
        entities,
      };
    })(intent.data))(acc);
  };

  const buildEntitySynonyms = (acc, _, name) => {
    return fp.append(fp.reduce((p, c) => {
      const synonyms = augment.getSynonyms(c);

      if (synonyms.length === 0) {
        return p;
      }

      return fp.append({
        value: c,
        synonyms,
      })(p);
    })(augment.getEntity(name).all()))(acc);
  };

  const buildRegexFeatures = (acc, _, name) => {
    const pattern = getRegexProp(getRealEntity(name));

    if (pattern) {
      return fp.append({
        name,
        pattern,
      })(acc);
    }

    return acc;
  };

  return utils.merge({
    rasa_nlu_data: {
      common_examples: fp.reduce(buildIntentExamples, [])(augment.getIntents()),
      regex_features: fp.reduce(buildRegexFeatures, [])(chatl.entities),
      lookup_tables: fp.reduce(buildLookupTable, [])(chatl.entities),
      entity_synonyms: fp.reduce(buildEntitySynonyms, [])(chatl.entities),
    }
  }, options);
}

File: python/pychatl/adapters/rasa.py
# pylint: disable=C0111

from pychatl import fp, utils
from pychatl.augment import Augment


def rasa(chatl, **options):
    """Convert a chatl dataset to a rasa representation as per
    https://rasa.com/docs/rasa/1.1.4/nlu/training-data-format/
    """
    augment = Augment(chatl, True)

    def get_real_entity(name):
        entity_type = augment.entities.get(name, {}).get('props', {}).get('type')

        if entity_type and entity_type in augment.entities:
            return entity_type

        return name

    def get_regex_prop(name):
        return augment.entities.get(name, {}).get('props', {}).get('regex')

    # For rasa, we need a map of synonyms -> value
    synonyms_lookup = fp.reduce(lambda acc, synonyms, value:
                                fp.append(
                                    *fp.map(lambda s: {s: value})(synonyms))(acc)
                                )(augment.synonyms_values)

    def build_lookup_table(acc, _, name):
        entity_name = get_real_entity(name)

        # Entity has regex feature, returns now
        if get_regex_prop(entity_name):
            return acc

        return fp.append({
            'name': name,
            'elements': augment.get_entity(entity_name).all(),
        })(acc)

    def build_intent_examples(acc, intent, name):
        def build_sentence(sentence):
            entities = []

            def reduce_sentence(result, cur):
                if not utils.is_entity(cur):
                    return result + cur.get('value')

                entity_name = get_real_entity(cur.get('value'))
                value = augment.get_entity(entity_name).next(cur.get('variant'))

                nonlocal entities

                entities.append({
                    'start': len(result),
                    'end': len(result) + len(value),
                    'entity': cur.get('value'),
                    # Check if its a synonym here
                    'value': synonyms_lookup.get(value, value),
                })

                return result + value

            return {
                'intent': name,
                'text': fp.reduce(reduce_sentence, '')(sentence),
                'entities': entities,
            }

        return fp.append(*fp.map(build_sentence)(intent.get('data', [])))(acc)

    def build_entity_synonyms(acc, _, name):
        def reduce_entity(result, cur):
            synonyms = augment.get_synonyms(cur)

            if not synonyms:
                return result

            return fp.append({
                'value': cur,
                'synonyms': synonyms,
            })(result)

        return fp.append(*fp.reduce(reduce_entity)(augment.get_entity(name).all()))(acc)

    def build_regex_features(acc, _, name):
        pattern = get_regex_prop(get_real_entity(name))

        if pattern:
            return fp.append({
                'name': name,
                'pattern': pattern,
            })(acc)

        return acc

    return utils.merge({
        'rasa_nlu_data': {
            'common_examples': fp.reduce(build_intent_examples, [])(augment.get_intents()),
            'regex_features': fp.reduce(build_regex_features, [])(augment.entities),
            'lookup_tables': fp.reduce(build_lookup_table, [])(augment.entities),
            'entity_synonyms': fp.reduce(build_entity_synonyms, [])(augment.entities),
        },
    }, options)
